lec03-basic-r.Rmd

```diff
@@ -1,4 +1,732 @@
+          the object)
+    * `nrow(surveys)` - returns the number of rows
+    * `ncol(surveys)` - returns the number of columns
+
+* Content:
+    * `head(surveys)` - shows the first 6 rows
+    * `tail(surveys)` - shows the last 6 rows
+
+* Names:
+    * `names(surveys)` - returns the column names (synonym of `colnames()` for `data.frame`
+	   objects)
+    * `rownames(surveys)` - returns the row names
+
+* Summary:
+    * `str(surveys)` - structure of the object and information about the class, length and
+	   content of  each column
+    * `summary(surveys)` - summary statistics for each column
+
+Note: most of these functions are "generic", they can be used on other types of
+objects besides `data.frame`.
+
+> ### Challenge
+>
+> Based on the output of `str(surveys)`, can you answer the following questions?
+>
+> * What is the class of the object `surveys`?
+> * How many rows and how many columns are in this object?
+> * How many species have been recorded during these surveys?
+
+<!---
+```{r, echo=FALSE}
+## Answers
+##
+## * class: data frame
+## * how many rows: 34786,  how many columns: 13
+## * how many species: 48
+```
+--->
+
+
+## Indexing and subsetting data frames
+
+Our survey data frame has rows and columns (it has 2 dimensions), if we want to
+extract some specific data from it, we need to specify the "coordinates" we want
+from it. Row numbers come first, followed by column numbers. However, note that
+different ways of specifying these coordinates lead to results with different
+classes.
+
+```{r}
+surveys[1, 1]   # first element in the first column of the data frame (as a vector)
+surveys[1, 6]   # first element in the 6th column (as a vector)
+surveys[, 1]    # first column in the data frame (as a vector)
+surveys[1]      # first column in the data frame (as a data.frame)
+surveys[1:3, 7] # first three elements in the 7th column (as a vector)
+surveys[3, ]    # the 3rd element for all columns (as a data.frame)
+head_surveys <- surveys[1:6, ] # equivalent to head(surveys)
+```
+
+`:` is a special function that creates numeric vectors of integers in increasing
+or decreasing order, test `1:10` and `10:1` for instance. This works similar to
+`seq`, which we looked at earlier in class:
+
+```{r}
+0:10
+seq(0, 10)
+
+# We can test if all elements are the same
+0:10 == seq(0,10)
+all(0:10 == seq(0,10))
+```
+
+You can also exclude certain parts of a data frame using the "`-`" sign:
+
+```{r}
+surveys[,-1]    # All columns, except the first
+surveys[-c(7:34786),] # Equivalent to head(surveys)
+```
+
+As well as using numeric values to subset a `data.frame` (or `matrix`), columns
+can be called by name, using one of the four following notations: <!-- Note sure how important it is to learn these difference vs just teaching the preferred way with the footnote that there are other ways also. -->
+
+```{r}
+head(surveys["species_id"])       # Result is a data.frame
+head(surveys[, "species_id"])     # Result is a vector
```

Comment:
This line is not correct. It results in a `data.frame`/`tbl_df`.

Reply:
Actually, this confused me a bit now that I went through it. I was expecting that I would get a `tbl_df` back every time, just as with the numerical indexing, but the two last indexing options actually returns a vector...

Reply:
Ah ok, I got part of it at least. I never included `[[]]` for the numerical indeces. Doing something like `species[[1]]` also returns a vector. I guess `$` just is shorthand for the `[[]]` type indexing since that also returns a vector?

Reply:
I think I will just leave out `[[]]` indexing... `$` I feel is common though, is that correct? I mean, with `dplyr` we use `select` anyways so I guess I can keep it simple here.

Reply:
Exactly. `$` is pretty common so I would talk about it. And yes `$` is short for `[[` since using `$` you can only select one column that way.